### PR TITLE
Add checksummed and new raja versions

### PR DESCRIPTION
**Pull request by gardner48**

### CODE REVIEW

File: var/spack/repos/builtin/packages/raja/package.py [hunk 1]
```diff
@@ -13,32 +13,42 @@ class Raja(CachedCMakePackage, CudaPackage, ROCmPackage):
...
+    version("0.9.0", sha256="b251f06676d88e2fa3ef30ccac04d55655318429a9ef1e008888b901593c5b3a")
+    version("0.8.0", sha256="c397b998bb3c847c6af03e44d85a2151b76a100bb2e883f28453f3689d90d394")
+    version("0.7.0", sha256="4127598d226936c0ea8d6e0b40da1049e65fce8efd0359bd391ec4ad7e4ecdcd")
+    version("0.6.0", sha256="d4435b5ff48443ea253d41df02b6d2988645eb79590cb3be1601664d4356e654")
+    version("0.5.3", sha256="f27d05ac77c13f4304967991539c34f946b05ee2d4628c2dcd9637c94b25e56f")
+    version("0.5.2", sha256="49d8ac6061ef3cda35c00c1d76ce5f52e4de49984863fcb28217331cf5f6a190")
+    version("0.5.1", sha256="af322bc890ed763db815d19ebe1fdb3c4c2b33855b10cdeb99b95a835f93b86e")
+    version("0.5.0", sha256="fde0a29a4f4cb35a66d9c0b34170b5bde2ceff8f8291d0c9166e9e0761012f22")
+    version("0.4.1", sha256="921d286b307215e2fa61c8424ba8039b145d3c4545bc7f12d9787e82e3bc8957")
+    version("0.4.0", sha256="ec90f91568dc9125b9150463259d069ea7ba52b38814aeb024e37708f27d2c43")
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/raja/package.py:51]: I cannot confirm these version sha256 as `spack checksum` reports:

```
==> Error: Could not find any remote versions for raja
```

**Review by tldahlgren — COMMENTED** [review]
I am able to confirm all but the last 8 version sha256.

File: var/spack/repos/builtin/packages/raja/package.py (hunk 1 above)
Comment (gardner48) [reply var/spack/repos/builtin/packages/raja/package.py:51]: Not sure why but I had to edit the `url` for each of these older versions to get `spack checksum` to find them e.g., `https://github.com/LLNL/RAJA/releases/download/v0.4.0/RAJA-0.4.0.tar.gz`

**Comment by adrienbernede** [comment]
@davidbeckingsale @rhornung67
If you agree with this, I’d like to see this merged to integrate it in RADIUSS Spack Configs and avoid conflicts when I push our local versions upstream.
Also: Is this something we want for all the RADIUSS packages ?

**Comment by rhornung67** [comment]
A couple of questions:

1. @gardner48 Why switch from using tags to sha hashes? Does it gain anything apart from being more verbose?
2. @adrienbernede we have newer releases for all RAJA Suite projects and updated Spack packages to reflect those in the radiuss-spack-configs repo. Why don't we put up a PR to update all of those?

**Comment by gardner48** [comment]
> 1. @gardner48 Why switch from using tags to sha hashes? Does it gain anything apart from being more verbose?

The proposed changes download the tarballs for the releases rather than using the tags so the downloads can be verified using a checksum (the `sha256` is the checksum hash rather than the commit hash). If using tags is easier for syncing with the radiuss-spack-configs repo that's good with me too. 

> 2. @adrienbernede we have newer releases for all RAJA Suite projects and updated Spack packages to reflect those in the radiuss-spack-configs repo. Why don't we put up a PR to update all of those?

If it's easy to setup, it might be handy to have a github action/bot to keep the package.py files up to date.

**Comment by adrienbernede** [comment]
@rhornung67 @gardner48 
I am preparing to push the local version of the RADIUSS packages indeed. However I am anticipating 2 things:
- Potential conflicts with existing PRs. Hence my questions about this one. I was wondering if sha hashes was the direction we were supposed to go.
- Potential push back due to the fact we share utility functions as part of the camp package. I asked the Spack team for feedback earlier this week, but I got nothing.

So, yes, I think I’m just going to submit a PR and have the discussion directly in there.

File: var/spack/repos/builtin/packages/raja/package.py (hunk 1 above)
Comment (tldahlgren) [reply var/spack/repos/builtin/packages/raja/package.py:51]: Spack automatically substitutes the version in the `url` property.  In this case the presence/absence of the `v` in the archive file is the issue in terms of URL substitution.

File: var/spack/repos/builtin/packages/raja/package.py [hunk 2]
```diff
@@ -13,32 +13,42 @@ class Raja(CachedCMakePackage, CudaPackage, ROCmPackage):
     """RAJA Parallel Framework."""
 
     homepage = "https://software.llnl.gov/RAJA/"
+    url = "https://github.com/LLNL/RAJA/releases/download/v2022.03.0/RAJA-v2022.03.0.tar.gz"
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/raja/package.py:16]: You could remove this property and add something along the lines of the following as the first method (after the dependencies, so around line 103):

```
    def url_for_version(self, version):
        vers = f"v{version}" if version > Version("0.8") else str(version)
        return f"https://github.com/LLNL/RAJA/releases/download/v{version}/RAJA-{vers}.tar.gz"
```

**Review by tldahlgren — CHANGES_REQUESTED** [review]
Still having issues with the older releases.  I have tested the suggested `url_for_version` method and it resolves the problem.

**Comment by tldahlgren** [comment]
> * Potential push back due to the fact we share utility functions as part of the camp package. I asked the Spack team for feedback earlier this week, but I got nothing.

Just sent a response with my opinion on this. Bottom line: there is precedence for adding a mix-in package that could help with all but one of the packages currently inheriting from `CachedCMakePackage` though it could be the first to focus on arguments whereas the current mix-ins are used to set the `url` and `mirrors`.

> 
> So, yes, I think I’m just going to submit a PR and have the discussion directly in there.

That's a great idea.

**Comment by gardner48** [comment]
Closing this PR in favor of #39613

**Comment by adrienbernede** [comment]
@gardner48, in #39613 I benefited from another PR that added the commits hash to the version tag. To me, this seems like a good security improvement. If you prefer tarball hashes for RAJA, please let me know.

**Comment by gardner48** [comment]
@adrienbernede I don't have a preference so whatever works best for the RAJA and RADIUSS teams is good with me

**Pull request closed without merging**